README.md
# NAME

CPAN::Tester::Box - A CPAN::Tester box based on [CPAN](https://metacpan.org/pod/CPAN).

# SYNOPSIS

```perl
use CPAN::Tester::RecentUploads;
use CPAN::Tester::Box;

my $recents = CPAN::Tester::RecentUploads->new(
    mirror => 'http://www.cpan.org',
);
my $box = CPAN::Tester::Box->new(
    recent_uploads => $recents,
    poll_interval  => $option{interval},
    tester_perl    => $option{perl},
    verbose        => 1,
);

$box->run();
```

# ATTRIBUTES

## recent\_uploads

_InstanceOf_ [CPAN::Tester::RecentUploads](https://metacpan.org/pod/CPAN%3A%3ATester%3A%3ARecentUploads)

_Required_

## poll\_interval

Interval between polls to the mirror for new files (in seconds).

_Default_: **3600** (1 hour)

## tester\_perl

The perl-binary to use for testing.

_Default_: **$^X**

## verbose

Determines the amount of output, **0 | 1 | 2**.

_Default_: **1**

## install\_tested

Call `CPAN::Shell->install_tested()` at the end.

_Default_: **0**

## o\_conf

A HashRef of CPAN configuration options to set for testing.

_Default_: **{ test\_report: 1 }**

## ignore

An ArrayRef of regex-patterns to test against the path:
`A/AU/AUTHORNAME/Dist-Thing-42.01.tgz`

_Default_: **\[ \]** (no ignore)

## skip\_initial

We start by collecting the recent uploads from the last week, if one wants to
skip these uploads, pass `skip_initial => 1` to the constructor.

_Default_: **0**

## handled

Administrative HashRef that keeps track of distributions handled.  This can be
initialised in order to "continue" where one left of.

## last\_poll

Administrative Int that keeps track of the last time we polled the mirror.

# DESCRIPTION

This is a very simple CPAN::Tester box. It polls the mirror for new files every
interval (3600 secs by default) and will call `CPAN::Shell->test('$tarball')` via `system()`. Whenever the queue of new files is empty, it will `sleep()`
for the remainder of the polling interval.

By calling `CPAN::Shell->test()`, we assume that you have configured the
CPAN module to create and send test reports (see [CPAN::Reporter](https://metacpan.org/pod/CPAN%3A%3AReporter)).

Advice: Run this programme as a separate user on the system, with its own CPAN
configuration. If you choose to install dependencies, use the [local::lib](https://metacpan.org/pod/local%3A%3Alib)
model (or set `PERL_MM_OPT=INST_BASE=/home/<user>/perl5lib` and `PERL_MB_OPT=--install_base /home/<user>/perl5lib` yourself).

**WARNING: By running this programme one agrees to run random code on ones
machine. This can be dangerous for that machine! Make sure the user running it,
can do as little harm as posible.**

# COPYRIGHT

© MMXXIII - Abe Timmerman <abeltje@cpan.org>

# LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


Changes
0.01_04 2023-11-05T22:36:17+01:00 (62e7b2b => Abe Timmerman)
 - (Abe Timmerman, Sun, 5 Nov 2023 22:36:17 +0100) Autocommit for
   distribution CPAN-Tester-Box 0.01_04 (same)

0.01_03 2023-11-05T15:47:42+01:00 (020a0a1 => Abe Timmerman)
 - (Abe Timmerman, Sun, 5 Nov 2023 15:47:42 +0100) Autocommit for
   distribution CPAN-Tester-Box 0.01_03 (test)
 - (Abe Timmerman, Sun, 5 Nov 2023 16:21:12 +0100) Regenerate README.md
 - (Abe Timmerman, Sun, 5 Nov 2023 18:35:38 +0100) Add the creation of a
   pid-file
 -     Easier to kill the process if you have the PID.
 -     We will also remove it on exit.
 - (Abe Timmerman, Sun, 5 Nov 2023 18:59:27 +0100) Add 'o_conf' a hashref
   with CPAN options
 -     One can now pass a hashref of CPAN options that need to be set
   before
 -     calling CPAN::Shell->test().
 -     These options can be set from the 'cpan-tester-box.yml' or from the
 -     command-line:
 -	 -o_conf test_report=1 --o_conf make_install_make_command=make
 - (Abe Timmerman, Sun, 5 Nov 2023 22:08:04 +0100) Add 'ignore' and
   'skip_initial'
 -     ignore: an arrayref of (regex) patterns that are each tested against
 -	 the path (A/AU/AUTHORNAME/Dist-Thing-42.01.tar.gz). The item will
   not
 -	 be tested if 1 or more patterns match.
 -	 This option can be set via 'cpan-tester-box.yml' or via command
   line:
 -	   -ignore '\bAUTHORNAME\b' --ignore 'Dist-Thing-(?=\d)'
 -     skip_initial: do not handle the initial items (1Week) but mark as
 -	 handled.
 -	 This option can be set via 'cpan-tester-box.yml' or via command
   line:
 -	   -skipinitial or use cpan-tester-box.yml

0.01_02 2023-11-05T14:17:12+01:00 (6e4f149 => Abe Timmerman)
 - (Abe Timmerman, Sun, 5 Nov 2023 14:17:12 +0100) Autocommit for
   distribution CPAN-Tester-Box 0.01_02 (test)
 - (Abe Timmerman, Sun, 5 Nov 2023 15:14:26 +0100) Introduce
   'install_tested'
 -     After running CPAN::Shell->test() we can also run
 -	 CPAN::Shell->install_tested() to install the tested modules. This
 -	 saves rebuilding and retesting if the module is a dependency for
   other
 -	 modules.
 -     Factor-out the actual test command, now we change it and it improves
 -	 testability.

0.01_01 2023-11-04T13:00:05+01:00 (7cf562b => Abe Timmerman)
 - (Abe Timmerman, Sat, 4 Nov 2023 13:00:05 +0100) Autocommit for
   distribution CPAN-Tester-Box 0.01_01 (test)
 - (Abe Timmerman, Sun, 5 Nov 2023 14:12:48 +0100) Use pipe-open to read
   from the ->test() command
 -     In order to make verbose=1 give some information about the reports,
   we
 -     now use a pipe-open to the command so we can evaluate it on the fly.
 -     Some module have so many dependencies that it takes a fair while
   before
 -     one can print captured output, we want some indication of progress
   so we
 -     need to process output on the fly.

0.00_00 2023-11-02T08:30:22+01:00 (6aecb2d => abeltje)
 - (abeltje, Thu, 2 Nov 2023 08:30:22 +0100) First commit of very bare
   software
 -     This will need some fine-tuning and a lot of POD.
 - (Abe Timmerman, Fri, 3 Nov 2023 21:15:42 +0100) It's starting to look
   like a module
 -     Add files for the distribution (Makefile.PL, MANIFEST.SKIP,
 -     README.md, Changes etc)
 -     Add t::Test::abeltje
 -     Add tests for the *.pm
 - (Abe Timmerman, Fri, 3 Nov 2023 21:32:46 +0100) Clean-up tests
 -     Remove 'redefined' warning
 -     t/20-cpan-tester-box.t: convert to t::Test::abeltje
 - (Abe Timmerman, Fri, 3 Nov 2023 22:22:45 +0100) Capture test output and
   test it.
 -     All output from CPAN::Tester::Box goes to STDERR, we capture it
   during
 -     the test-run and examine it for some things we expect.
 - (Abe Timmerman, Fri, 3 Nov 2023 22:40:29 +0100) Add 'xtest' target to
   the Makefile
 - (Abe Timmerman, Sat, 4 Nov 2023 09:56:39 +0100) Introduce 'verbose' to
   the Box
 -     Now one can determine the amount of output:
 -	 0 => none
 -	 1 => some administrative messages
 -	 2 => also the output of CPAN::Shell->test()
 -     Remove Capture::Tiny
 - (Abe Timmerman, Sat, 4 Nov 2023 10:08:52 +0100) Exclude some programme
   files
 -     cpan-tester-box.db
 -     cpan-tester-box.yml
 - (Abe Timmerman, Sat, 4 Nov 2023 11:32:58 +0100) Add 'EXE_FILES' to
   WriteMakefile
 -     We want the script to be installed.
 -     regen META.*


Makefile.PL
#! perl -w
use strict;
use v5.20.0;

use ExtUtils::MakeMaker::CPANfile;

WriteMakefile(
    NAME             => 'CPAN::Tester::Box',
    VERSION_FROM     => 'lib/CPAN/Tester/Box.pm',
    ABSTRACT_FROM    => 'lib/CPAN/Tester/Box.pm',
    AUTHOR           => 'Abe Timmerman <abeltje@cpan.og>',
    LICENSE          => 'perl_5',
    MIN_PERL_VERSION => 5.20.0,

    # Dependencies managed in cpanfile
    EXE_FILES => [qw< bin/cpan-tester-box >],

    META_MERGE => {
        'meta-spec' => { version => 2 },
        'resources' => {
            repository => {
                type => 'git',
                url  => 'https://github.com/abeltje/CPAN-Test-Box.git',
                web  => 'https://github.com/abeltje/CPAN-Test-Box',
            },
            bugtracker => {
                web => 'https://github.com/abeltje/CPAN-Test-Box/issues',
            },
        },
    },
);

package MY;

sub test {
    my $self = shift;
    my $orig = $self->SUPER::test(@_);

    $orig .= <<"EOM";

XTEST_FILES = xt/*.t
xtest ::
EOM

    my $varpfx = $^O eq 'MSWin32' ? '$$' : '\$$';
    $orig .= "\t"
           . qq{\$(FULLPERLRUN) "-e" "my ${varpfx}sw=q[-wl]; }
           . qq{${varpfx}sw .= q[v] if \$(TEST_VERBOSE); }
           . qq{system qq[prove ${varpfx}sw \$(XTEST_FILES)]"};

    return $orig;
}
1;


lib/CPAN/Tester/RecentUploads.pm
package CPAN::Tester::RecentUploads;
use Moo;
use v5.20.0;
use if $] <  5.036, experimental => 'signatures';
use if $] >= 5.036, feature      => 'signatures';
use Carp;

our $VERSION = '0.01';

use CPAN::Recent::Uploads::Retriever;
use DateTime;
use YAML::XS qw< Load >;

=head1 NAME

CPAN::Testers::RecentUploads - Wrapper around L<CPAN::Recent::Uploads::Retriever>.

=head1 ATTRIBUTES

=head2 mirror

This is the URI for the mirror to use to determine recent uploads. This is an
B<InstanceOf> either L<URI::http> or L<URI::https>, but strings are coerced into
a L<URI> instance.

I<Default>: B<http://www.cpan.org>

=head2 interval

This is one of: B<1h>, B<6h>, B<1d>, B<1W>, B<1M>, B<1Q> or B<1Y>. See
L<CPAN::Recent::Uploads> for details.

I<Default>: B<1W>

=cut

use URI;
use Types::Standard qw< Enum InstanceOf >;
has mirror => (
    is      => 'ro',
    isa     => InstanceOf([ 'URI::http', 'URI::https' ]),
    coerce  => sub { return URI->new($_[0]); },
    default => 'http://www.cpan.org'
);
has interval => (
    is      => 'ro',
    isa     => Enum [qw< 1h 6h 1d 1W 1M 1Q 1Y >],
    default => '1W'
);

my $_ext_re = qr{ \. (?: tar\.gz | tar\.bz2 | tgz | zip ) }x;

=head1 SYNOPSIS

    my $recents = CPAN::Tester::RecentUploads->new(
        mirror => 'https://www.cpan.org',
    );

=head1 DESCRIPTION

=head2 $ru->get_recent($interval)

=cut

sub get_recent ($self, $interval) {
    $interval //= $self->interval;

    my $yaml = CPAN::Recent::Uploads::Retriever->retrieve(
        time   => $interval,
        mirror => $self->mirror->as_string,
    );
    my $recent;
    eval { $recent = Load($yaml); 1 } or confess("Cannot unyaml: $!");

    my $list = [ map {
        (my $path = $_->{path}) =~ s{^ id/ }{}x;
        {
            path  => $path,
            epoch => $_->{epoch},
            time  => DateTime->from_epoch(
                epoch => $_->{epoch},
                ($ENV{TZ} ? (time_zone => $ENV{TZ}) : ()),
            ),
        }
    }sort {
        $a->{epoch} <=> $b->{epoch}
    } grep {
            $_->{path} =~ m{^ id/[A-Z]/[A-Z]{2}/ .+ $_ext_re $}x
        and $_->{type} eq 'new'
        and $_->{path} !~ m{ /perl-5\. }x
        and $_->{path} !~ m{ /perl6 }xi
    } @{$recent->{recent}} ];

    return $list;
}

use namespace::autoclean;
1;

=head1 COPYRIGHT

E<copy> MMXXIII - Abe Timmerman <abeltje@cpan.org>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

=cut


lib/CPAN/Tester/Box.pm
package CPAN::Tester::Box;
use Moo;
use v5.20.0;
use if $] <  5.036, experimental => 'signatures';
use if $] >= 5.036, feature      => 'signatures';

our $VERSION = "0.01_04";

use File::Spec::Functions qw< devnull >;
use DateTime;

=head1 NAME

CPAN::Tester::Box - A CPAN::Tester box based on L<CPAN>.

=head1 SYNOPSIS

    use CPAN::Tester::RecentUploads;
    use CPAN::Tester::Box;

    my $recents = CPAN::Tester::RecentUploads->new(
        mirror => 'http://www.cpan.org',
    );
    my $box = CPAN::Tester::Box->new(
        recent_uploads => $recents,
        poll_interval  => $option{interval},
        tester_perl    => $option{perl},
        verbose        => 1,
    );

    $box->run();

=head1 ATTRIBUTES

=head2 recent_uploads

I<InstanceOf> L<CPAN::Tester::RecentUploads>

I<Required>

=head2 poll_interval

Interval between polls to the mirror for new files (in seconds).

I<Default>: B<3600> (1 hour)

=head2 tester_perl

The perl-binary to use for testing.

I<Default>: B<$^X>

=head2 verbose

Determines the amount of output, B<0 | 1 | 2>.

I<Default>: B<1>

=head2 install_tested

Call C<< CPAN::Shell->install_tested() >> at the end.

I<Default>: B<0>

=head2 o_conf

A HashRef of CPAN configuration options to set for testing.

I<Default>: B<{ test_report: 1 }>

=head2 ignore

An ArrayRef of regex-patterns to test against the path:
C<A/AU/AUTHORNAME/Dist-Thing-42.01.tgz>

I<Default>: B<[ ]> (no ignore)

=head2 skip_initial

We start by collecting the recent uploads from the last week, if one wants to
skip these uploads, pass C<< skip_initial => 1 >> to the constructor.

I<Default>: B<0>

=head2 handled

Administrative HashRef that keeps track of distributions handled.  This can be
initialised in order to "continue" where one left of.

=head2 last_poll

Administrative Int that keeps track of the last time we polled the mirror.

=cut

use Types::Standard qw< ArrayRef Bool HashRef InstanceOf Int Str >;
has recent_uploads => (
    is       => 'ro',
    isa      => InstanceOf ['CPAN::Tester::RecentUploads'],
    required => 1,
);
has handled => (
    is      => 'ro',
    isa     => HashRef,
    default => sub { {} },
);
has poll_interval => (
    is      => 'ro',
    isa     => Int,
    default => sub { 60 * 60 },
);
has tester_perl => (
    is      => 'ro',
    isa     => Str,
    default => sub {$^X},
);
has verbose => (
    is      => 'ro',
    isa     => Int,
    default => 1,
);
has install_tested => (
    is      => 'ro',
    isa     => Bool,
    default => 0,
);
has o_conf => (
    is      => 'ro',
    isa     => HashRef,
    default => sub { { test_report => 1 } },
);
has ignore => (
    is      => 'ro',
    isa     => ArrayRef,
    default => sub { [] },
);
has skip_initial => (
    is      => 'ro',
    isa     => Bool,
    default => 0,
);
has last_poll => (
    is  => 'rw',
    isa => Int,
);

=head1 DESCRIPTION

This is a very simple CPAN::Tester box. It polls the mirror for new files every
interval (3600 secs by default) and will call C<< CPAN::Shell->test('$tarball')
>> via C<system()>. Whenever the queue of new files is empty, it will C<sleep()>
for the remainder of the polling interval.

By calling C<< CPAN::Shell->test() >>, we assume that you have configured the
CPAN module to create and send test reports (see L<CPAN::Reporter>).

Advice: Run this programme as a separate user on the system, with its own CPAN
configuration. If you choose to install dependencies, use the L<local::lib>
model (or set C<< PERL_MM_OPT=INST_BASE=/home/<user>/perl5lib >> and C<<
PERL_MB_OPT=--install_base /home/<user>/perl5lib >> yourself).

B<WARNING: By running this programme one agrees to run random code on ones
machine. This can be dangerous for that machine! Make sure the user running it,
can do as little harm as posible.>

=begin documentation

=head2 $box->get_recent_queue($interval)

Returns an ArrayRef with items from the last C<$interval> but without those that
are in C<< $self->handled >>.

=cut

sub get_recent_queue ($self, $interval) {
    $self->last_poll(time());

    my $all_recent = $self->recent_uploads->get_recent($interval) // [ ];
    if ($self->verbose and @$all_recent) {
        my $last = $all_recent->[-1];
        say STDERR "# Last($interval): $last->{path} $last->{time}"
    }

    my $recent = [ grep {
        ! exists($self->handled->{$_->{path}})
    } @$all_recent ];
    return $recent;
}

=head2 $box->wants_ignore($item)

=cut

sub wants_ignore ($self, $item) {
    my @ignore = grep {
        $item->{path} =~ m{$_}
    } @{ $self->ignore };
    return @ignore ? [ @ignore ] : ();
}

=head2 $self->_test_command($item)

This returns an ArrayRef with the elements of the command to run.

mostly to improve testability...

=cut

sub _test_command ($self, $item) {
    my $install_tested = $self->install_tested
        ? 'CPAN::Shell->install_tested();'
        : "";

    my @o_conf = map {
        "CPAN::Shell->o(conf => '$_', '@{[ $self->o_conf->{$_} ]}');"
    } sort keys %{ $self->o_conf };

    my $indent = " " x 12;
    my @cmd = (
        sprintf('"%s"', $self->tester_perl),
        '"-MCPAN"',
        '"-e"',
        qq/"
            CPAN::HandleConfig->require_myconfig_or_config;
            @{[ join("\n$indent", @o_conf) ]}
            CPAN::Shell->test('$item->{path}');
            $install_tested
        "/
    );
    return \@cmd;
}

=head2 $box->handle_queue_item($item)

    $self->tester_per -MCPAN -e "CPAN::Shell->test('$item->{path}')"

=cut

sub handle_queue_item ($self, $item) {
    return if !$item;
    return if $self->handled->{ $item->{path} };
    if (my $sp = $self->wants_ignore($item)) {
        say STDERR "# IGNORE: $item->path $item->{time}" if $self->verbose;
        say STDERR "# IPAT: @$sp" if $self->verbose;
        $self->handled->{ $item->{path} }++;
        return 1;
    }
    say STDERR "# $item->{path} $item->{time}" if $self->verbose;

    my $cmd = $self->_test_command($item);
    my $cmdln = join(" ", @$cmd);
    say STDERR "# $cmdln" if $self->verbose;

    {
        local $ENV{AUTOMATED_TESTING} = 1;
        local $ENV{PERL_MM_USE_DEFAULT} = 1;
        if (open(my $td, '-|', "$cmdln 2>&1")) {
            while (my $line = <$td>) {
                if ($self->verbose > 1) {
                    print STDERR $line;
                }
                elsif ($self->verbose) {
                    print STDERR "# $line" if $line =~ m{^CPAN::Reporter:};
                }
            }
            close($td) or say STDERR "# Error on close-pipe: $!";
        }
        else {
            say STDERR "# Couldn't open-pipe: $! ($?)";
            my $redir = $self->verbose ? '1>&2' : '2>&1 >' . devnull;
            system("$cmdln $redir");
        }
    }

    $self->handled->{ $item->{path} }++;
    return 1;
}

=head2 $box->top_up_queue($queue)

If we are at or over the C<< $self->poll_interval >> time, just push new items
on the queue.

=cut

sub top_up_queue ($self, $queue) {
    my $interval = time() - $self->last_poll - $self->poll_interval;
    push(@$queue, @{$self->get_recent_queue('6h')}) if $interval >= 0;
    return 1;
}

=head2 $box->handle_queue($queue)

While there are queue-items, I<handle-item> and I<top-up-queue>

=cut

sub handle_queue ($self, $queue) {
    while (@$queue) {
        my $item = shift(@$queue);
        $self->handle_queue_item($item);
        $self->top_up_queue($queue);
    }
}

=head2 $box->wait_for_new($queue)

Sleep until the next C<< $self->poll_intervall >> and I<top-up-queue>.

=cut

sub wait_for_new ($self, $queue) {
    my $interval = $self->poll_interval - (time() - $self->last_poll);
    my $now = DateTime->now(); $now->set_time_zone($ENV{TZ}) if $ENV{TZ};
    say STDERR sprintf(
        "# wait_for_new: sleep(%u) of %u %s",
        $interval, $self->poll_interval, $now->strftime("%a %F %T %z")
    ) if $self->verbose;

    sleep($interval) if $interval > 0;
    return $self->top_up_queue($queue);
}

=head2 $box->run

Fetch recent items (1W) and endlessly repeat I<handle-queue>, I<wait-for-new>.

=cut

sub run ($self) {
    my $queue = $self->get_recent_queue('1W');
    if ($self->skip_initial) {
        while (@$queue) {
            my $item = shift(@$queue);
            say STDERR "# SKIP $item->{path} $item->{time}" if $self->verbose;
            $self->handled->{ $item->{path} }++;
        }
    }
    while (1) {
        $self->handle_queue($queue);
        $self->wait_for_new($queue);
    }
}

use namespace::autoclean;
1;

=end documentation

=head1 COPYRIGHT

E<copy> MMXXIII - Abe Timmerman <abeltje@cpan.org>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

=cut
